incubator/hnc/controllers/object_controller.go

```diff
@@ -102,63 +104,110 @@ func (r *ObjectReconciler) SyncNamespace(ctx context.Context, log logr.Logger, n
-		if yes, err := r.isAncestor(inst.GetNamespace(), inheritedFrom); err != nil {
-			// This can fail if the names are bad. For now, take no action (ie don't
-			// delete). TODO: revisit this.
-			log.Error(err, "deleteObsolete")
-			return false, err
-		} else if yes {
-			// We found the object in an ancestor namespace; no need to delete it.
-			return false, nil
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("The source object no longer exists", "inheritedFrom", srcNS)
+			return true, nil, nil
+		} else {
+			log.Info("Couldn't read", "source", srcNm)
+			return false, nil, err
 		}
-		log.Info("Will delete because the source namespace is no longer an ancestor", "inheritedFrom", inheritedFrom)
-	} else if errors.IsNotFound(err) {
-		log.Info("Will delete because the source object no longer exists", "inheritedFrom", inheritedFrom)
-	} else {
-		log.Info("Couldn't read", "source", srcNm)
-		return false, err
 	}
 
-	// Delete this instance.
-	if err := r.Delete(ctx, inst); err != nil {
-		log.Error(err, "Couldn't delete after source was missing")
+	// The source exists. Is it still in an ancestor namespace?
+	if yes, err := r.isAncestor(inst.GetNamespace(), srcNS); err != nil {
+		// This can fail if the names are bad. For now, take no action (ie don't
+		// delete). TODO: revisit this.
+		log.Error(err, "deleteObsolete")
+		return false, nil, err
+	} else if !yes {
+		log.Info("The source namespace is no longer an ancestor", "inheritedFrom", srcNS)
+		return true, nil, nil
+	}
+
+	return false, src, nil
+}
+
+// ensureCorrectAncestry checks to see if the given object is correct. It returns true if
+// it's correct and ready to be propagated. The incorrect object will be deleted
+// if it's obsolete; or marked as modified if it's changed.
+func (r *ObjectReconciler) ensureCorrectAncestry(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, error) {
+	log.Info("Ensure correctness")
+	// Find what we expect the source to be.
+	missing, src, err := r.getSourceInst(ctx, log, inst)
+	if err != nil {
 		return false, err
 	}
 
-	return true, nil
+	// The source instance no longer exists in the source namespace.
+	if missing {
+		log.Info("Will delete because the source object no longer exists in ancestor namespace(s)", "inheritedFrom", inst.GetLabels()[labelInheritedFrom])
+		return false, r.Delete(ctx, inst)
+	}
+
+	// If there's no source, *this* is the source
+	if src == nil {
+		return true, nil
+	}
+
+	// We found the object in an ancestor namespace. Is this object the same as the source?
+	log.Info("Compare with the source", "inheritedFrom", src.GetNamespace())
+	instCopied := copyObject(inst)
+	// Make the namespaces and inheritedFrom labels the same before comparison.
+	instCopied.SetNamespace(src.GetNamespace())
+	setLabel(instCopied, labelInheritedFrom, "")
+	srcCopied := copyObject(src)
+	setLabel(srcCopied, labelInheritedFrom, "")
+	equal := reflect.DeepEqual(instCopied, srcCopied)
```

Comment:
Please review the comparison:

The only possible known differences are 1) namespace and 2) "labelInheritedFrom", so I changed the code to reset these 2 fields before comparison and had a comment above it.
Here, I only reset the labelInheritedFrom label instead of the label map[string]string in case anyone modified the labels. Is it necessary?